opentracing-api/src/main/java/io/opentracing/ActiveSpan.java

```diff
@@ -0,0 +1,71 @@
+package io.opentracing;
+
+import java.io.Closeable;
+
+/**
+ * In any execution context (or any thread, etc), there is at most one "active" {@link Span} primarily responsible for
+ * the work accomplished by the surrounding application code. That active Span may be accessed via the
+ * {@link ActiveSpanSource#active()} method. If the application needs to defer work that should be part of the same Span, the
+ * Source provides a {@link ActiveSpan#defer} method that returns a {@link Continuation}; this continuation may be used
+ * to re-activate and continue the {@link Span} in that other asynchronous executor and/or thread.
+ *
+ * <p>
+ * {@link ActiveSpan}s are created via {@link Tracer.SpanBuilder#startAndActivate()} or {@link ActiveSpanSource#adopt}. They can
+ * be {@link ActiveSpan#defer()}ed as {@link ActiveSpan.Continuation}s, then re-{@link Continuation#activate()}d later.
+ *
+ * @see ActiveSpanSource
+ */
+public interface ActiveSpan extends Closeable, Span {
+    /**
+     * Mark the end of the active period for the {@link Span} pinned by this {@link ActiveSpan}. When the last
+     * {@link ActiveSpan} is deactivated for a given {@link Span}, it is automatically {@link Span#finish()}ed.
+     * <p>
+     * <p>
+     * NOTE: It is an error to call deactivate() more than once on a single {@link ActiveSpan}.
+     *
+     * @see Closeable#close() {@link ActiveSpan}s are auto-closeable and may be used in try-with-resources blocks
+     */
+    void deactivate();
```

Comment:
> in framework integrations start and finish usually run in separate threads...

Of course you are correct about this. IMO it means that the framework would need an actual (internal-to-framework) code change to make tracing sensible, at least when a threadlocal is involved. Not a great answer, but I don't see how we solve this with any approach that's optimized for threadlocals.

As for the refcounting: I am pretty certain that the increment needs to happen at `capture()`-time... But maybe we should change the contract for `finish()` to make it explicitly safe (but a noop) to call `finish()` twice on a single Span instance... that way the exception handler in your example could `finish()` the Span (and tag it with an error, etc) even if there are `Continuations` outstanding. Another important point would be to have something like `Continuation.cancel()` which would be a synonym for `Continuation.activate().deactivate()`. (I.e., it decrements the refcount)

Thoughts??